firewood/src/merkle.rs

```diff
@@ -1175,6 +1180,180 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
+                loop {
+                    match next {
+                        None => return Poll::Ready(None),
+                        Some((parent, child_position)) => {
+                            // This code assumes all parents are branch nodes and will panic otherwise
+                            if child_position as usize == NBRANCH - 1 {
+                                // don't index past end of list
+                                next = pinned_self.parents.pop();
+                                continue;
+                            }
+                            match parent.inner().as_branch().unwrap().chd()
+                                [child_position.wrapping_add(1) as usize]
+                            {
+                                Some(addr) => {
+                                    // there is a child at the next address, so walk down it
+                                    // We use u8::MAX to walk down the leftmost value, as it
+                                    // gets increased immediately once popped
+                                    let child = pinned_self
+                                        .merkle
+                                        .get_node(addr)
+                                        .map(|node| (node, u8::MAX))
+                                        .map_err(|e| api::Error::InternalError(e.into()))?;
+                                    let keep_going = child.0.inner().is_branch();
+                                    next = Some(child);
+                                    pinned_self
+                                        .parents
+                                        .push((parent, child_position.wrapping_add(1)));
+                                    if !keep_going {
+                                        break;
+                                    }
+                                }
+                                None => {
+                                    next = pinned_self.parents.pop();
+                                    continue;
+                                }
+                            }
+                        }
+                    }
+                }
+                // recompute current_key
+                // TODO: Can we keep current_key updated as we walk the tree instead of building it from the top all the time?
+                let current_key = pinned_self
+                    .parents
+                    .iter()
+                    .skip(1)
+                    .map(|parent| parent.1)
+                    .tuples()
+                    .map(|(hi, lo)| (hi << 4) + lo)
+                    .collect::<Vec<u8>>();
+                pinned_self.current_key = current_key.into(); // Some(current_node.inner().as_leaf().unwrap().0.to_vec());
+                pinned_self
+                    .current_key
+                    .as_mut()
+                    .unwrap()
+                    .extend(current_node.inner().as_leaf().unwrap().0.to_vec());
```

Comment:
You already have access to the leaf here in this part of the match arm, why not `_leaf` -> `leaf` and use `leaf.0.to_vec()`?